compiler/rustc_mir_dataflow/src/impls/single_enum_variant.rs

```diff
@@ -0,0 +1,171 @@
+        &self,
+        state: &mut Self::Domain,
+        statement: &Statement<'tcx>,
+        loc: Location,
+    ) {
+        let (place, fact) = match &statement.kind {
+            StatementKind::Deinit(box place) => (place, None),
+            StatementKind::SetDiscriminant { box place, variant_index } => {
+                (place, Some(*variant_index))
+            }
+            StatementKind::Assign(box (lhs, Rvalue::Use(op))) => {
+                return self.assign(state, lhs, op, loc);
+            }
+            /* may alias/mutate RHS need to specify that it is no longer a single value */
+            StatementKind::Assign(box (
+                _,
+                Rvalue::Ref(_, BorrowKind::Mut { .. }, rhs)
+                | Rvalue::AddressOf(Mutability::Mut, rhs),
+            )) => (rhs, None),
+            StatementKind::CopyNonOverlapping(box ref copy) => {
+                let place = match &copy.dst {
+                    Operand::Copy(p) | Operand::Move(p) => p,
+                    _ => return,
+                };
+                (place, None)
+            }
+            _ => return,
+        };
+        if !self.is_tracked(place) {
+            return;
+        }
+        let Some(local) = place.local_or_deref_local() else { return };
+        if let Some(fact) = fact {
+            state.insert(local, loc, fact);
+        } else {
+            state.remove(local);
+        }
+    }
+    fn apply_terminator_effect(
+        &self,
+        state: &mut Self::Domain,
+        terminator: &Terminator<'tcx>,
+        loc: Location,
+    ) {
+        match &terminator.kind {
+            TerminatorKind::DropAndReplace { place, value, .. } => {
+                self.assign(state, place, value, loc)
+            }
+            TerminatorKind::Drop { place, .. } if self.is_tracked(place) => {
+                let Some(local) = place.local_or_deref_local() else { return };
+                state.remove(local);
+            }
+            _ => {}
+        }
+    }
+
+    fn apply_call_return_effect(
+        &self,
+        _: &mut Self::Domain,
+        _: BasicBlock,
+        _: CallReturnPlaces<'_, 'tcx>,
+    ) {
+    }
+
+    fn apply_switch_int_edge_effects(
+        &self,
+        _block: BasicBlock,
+        discr: &Operand<'tcx>,
+        apply_edge_effects: &mut impl SwitchIntEdgeEffects<Self::Domain>,
+    ) {
+        let Some(place) = discr.place() else { return };
+        if !self.is_tracked(&place) {
+            return;
+        }
+        let Some(local) = place.local_or_deref_local() else { return };
+        apply_edge_effects.apply(|state, target| {
+            // This probably isn't right, need to check that it fits.
+            let new_fact = target.value.map(|v| VariantIdx::from_u32(v as u32));
+
+            if let Some(new_fact) = new_fact {
+                let loc = Location { block: target.target, statement_index: 0 };
+                state.insert(local, loc, new_fact);
```

Comment:
I think things are slightly mixed up here. In MIR like this:

```rust
_5 = Discrim(_8);
switchInt(_5, 0 => bb3, otherwise: bb5)
```

Your `local` will store `_5`, but really you want to insert a new fact about `_8`. Take a look at <https://doc.rust-lang.org/nightly/nightly-rustc/rustc_mir_dataflow/impls/fn.switch_on_enum_discriminant.html> for inspiration on how to do this kind of thing. Essentially, you will just want to check if the previous statement was an assignment of the shape given above.